_specifications/lsp/3.18/language/completion.md

```diff
@@ -386,6 +399,75 @@ export interface CompletionList {
 		data?: LSPAny;
 	}
 
+	/**
+	 * Specifies how fields from a completion item should be combined with those
+	 * from `completionList.itemDefaults`.
+	 * 
+	 * In unspecified, all fields will be treated as "replace".
+	 * 
+	 * If a field's value is "replace", the value from a completion item will
+	 * always be used instead of the value from `completionItem.itemDefaults`.
+	 * 
+	 * If a field's value is "merge", the values will be merged using the rules
+	 * defined against each field below.
+	 *
+	 * Servers are only allowed to return `applyKind` if the client
+	 * signals support for this via the `completionList.applyKinds`
+	 * capability.
+	 *
+	 * @since 3.18.0
+	 */
+	applyKind?: {
+		/**
+		 * Specifies whether commitCharacters on a completion will replace or be
+		 * merged with those in `completionList.itemDefaults.commitCharacters`.
+		 * 
+		 * If "replace", the commit characters from the completion item will
+		 * always be used unless not provided, in which case those from
+		 * `completionList.itemDefaults.commitCharacters` will be used. An empty
+		 * list can be used if a completion item does not have any commit
+		 * characters and also should not use those from
+		 * `completionList.itemDefaults.commitCharacters`.
+		 * 
+		 * If "merge" the commitCharacters for the completion will be the union
+		 * of all values in both `completionList.itemDefaults.commitCharacters`
+		 * and the completion's own `commitCharacters`. An empty list indicates
+		 * the completion adds no additional commit characters and a value of
+		 * `null` indicates the defaults should not be used.
+		 *
+		 * @since 3.18.0
+		 */
+		commitCharacters?: "replace" | "merge";
+
+		/**
+		 * Specifies whether data on a completion will replace or
+		 * be merged with data from `completionList.itemDefaults.data`.
+		 * 
+		 * If "replace", the data from the completion item will be used if
+		 * provided, otherwise  `completionList.itemDefaults.data` will be used.
+		 * An empty object can be used if a completion item does not have any
+		 * data but also should not use the value from
+		 * `completionList.itemDefaults.data`.
+		 * 
+		 * If "merge", a shallow merge will be performed between
+		 * `completionList.itemDefaults.data` and the completion's own data
+		 * using the following rules:
+		 * 
+		 * - If a completion's data is not specified, the data from
+		 *   `completionList.itemDefaults.data` will be used.
+		 * - If a completion's data is null, the resulting data field will be
+		 *   null (defaults are not used).
+		 * - If a field is specified in `completion.data` it will be used as-is.
+		 * - If a field is `null` in `completion.data`, it will remain `null`
+		 *   (default is not used).
```

Comment:
Isn't this the same point again?

Reply:
This one is about fields within `data`, not `data` itself:

```js
{
  label: 'MyCompletion',
  data: null,
}
```

vs

```js
{
  label: 'MyCompletion',
  data: {
    foo: null,
  },
}
```

Again, the intention was to allow `itemDefaults` to cover the majority case, but allow opting out for a subset of completion items. It's probably not critical because you could always work around this by having a boolean flag in defaults that is `true` and set to `false` on the completion item indicating to not use some other set of data.

I'll look at simplifying this too.